### ISSUE TITLE
Railcore config repo reorganization

### ISSUE THREAD

**Issue by lostapathy** [railcore/configs#9]
We've discussed various takes on this in IRC and in Facebook.  This issue is an attempt at actually getting the ball rolling.

We will move the slicer configs into a separate repository from the duet configuration files for the machine itself.  This will allow us to manage them in different ways.  

Mainly, there is a desire to streamline and unify the duet config to reduce duplication and make onboarding users easier.  Because they directly impact the machine configuration, we need to maintain a high level of quality control and keep safe/sane defaults there.  At the same time, there's a desire to make the slicer configs more easy going and encourage contributions in whatever form. 

Logically, slicer configs and duet configs are totally different animals and should be packaged separately anyway.

We will end up with separate repos as:

1. railcore/duet - which will work toward a single, unified duet configuration that is easily configurable for all machine types.  Changes to this repo should go through a review process to ensure they are "safe" and universally applicable. 
1. railcore/klipper - we will defer organization and processes on this to klipper users
1. railcore/slicers - this will be a place to store contributed slicer configs as-submitted and with little restriction.  Submissions will not need to be reviewed in detail.  User submissions should, however, be namespaced in a directory for that user and include a README that gives an overview of the intent of the slicer config.

Longer term we will probably want to make more detailed contributing guidelines for each repo, but I'm hoping this will be a good start to get us unstuck from where we are today.

**Comment by dandancheeseandham** [comment]
I agree with this on the whole, although I would add one proviso, and it stems from

> Because they directly impact the machine configuration, we need to maintain a high level of quality control and keep safe/sane defaults there. At the same time, there's a desire to make the slicer configs more easy going and encourage contributions in whatever form.

There should absolutely be a "release" level with the Duet config for all the excellent reasons you mention, but I believe that to encourage the same contribution mentality and a faster release/response schedule, there should be an "unsafe" area for Duet configs so people are able to share without having to jump through hoops (or resort to using Facebook/GDrive/Dropbox)

**Comment by natewalck** [comment]
For unsafe configs, I think we should use branches for that. Once they are tested, they would get merged into master.

**Comment by JohnOCFII** [comment]
Great ideas.  I do think that regardless of safe or not, there will be philosophical differences regarding what goes into the “master” released Duet RRF Configs.  So I think there should still be an area to share configs that may not be accepted as part of the master config.

**Comment by lostapathy** [comment]
Right - the expectation would be to use a normal development workflow for most changes.  Fork the repo, do your work there in a branch, then submit a PR for changes that should go upstream.

Perhaps, though, it would be good to have a listing with links to forks that are maintained (or at least, kept in a working state) that have an alternate philosophy that makes them interesting to the community, but not necessarily suitable to be upstreamed.

Depending on how things go, we could have branches in the main repo for "community ready" alternate configurations, but I feel like that's something we should hammer out after we get the main repo cleaned up and organized how we want it, so that all branches on the main repo are at least structured the same way.

**Comment by lostapathy** [comment]
As a side note, once this is settled we need to create a CONTRIBUTING.md in each repo that captures the results of this discussion (obviously subject to future revision).

We also need to add a license file to each repo at this time - I suggest MIT unless @kraegar has an issue with that.

**Comment by natewalck** [comment]
To @JohnOCFII's point, I think instead of having a user-duet-configs repo, each user can keep their configs in their own repo (like https://github.com/natewalck/railcoreiiduet/) and then we can list these repos in the README.md (or another file as user configurations that folks might want to  check out.

This main "official" repo should be very opinionated configurations that are generally safe and have default configs for each Z probe type, endstops, etc.

**Comment by dandancheeseandham** [comment]
OK, so what's next? 

> 1. railcore/duet - which will work toward a single, unified duet configuration that is easily configurable for all machine types.

Shall we start with the ProjectR3D folder, delete the others and get this train rolling?
Do we tag it with firmware versions or anything?

@kraegar indicated to go with the Thingiverse license, so as the Thingiverse models are under https://creativecommons.org/licenses/by/3.0/ I guess we go with that.

Also I'm very new to this Git/Github and development methodology in general, yet currently seem to have full access to this repo, so feel please free to restrict my rights to prevent me from bypassing any process that should be followed.

**Comment by lostapathy** [comment]
I think first step is to split the repo into 3 repos before we start changing the contents.  I believe @natewalck thought he had a handle on a good strategy for sorting out the Duet configs, so I will let him take the lead on that.

**Comment by plc268** [comment]
Adding on, I think it would be useful to collect and organize all the STLs and cad files and post them in one source. There's a whole host of files that are strewn around in separate places and are getting lost in the fold. Not to mention, that thingiverse hasn't exactly been the most reliable host lately.

**Comment by kraegar** [comment]
My plan is that once the wiki is on github I'll make a section for things
like the halo. Describe its function, how to install it, and link the
fusion model for it all in one spot.  Been holding off until the wiki moves.

**Comment by dandancheeseandham** [comment]
> Adding on, I think it would be useful to collect and organize all the STLs and cad files and post them in one source. There's a whole host of files that are strewn around in separate places and are getting lost in the fold. Not to mention, that thingiverse hasn't exactly been the most reliable host lately.

Are STL's config or BOM? I figured BOM as it becomes a material :) Anyway I've tried to do that, and it's currently in the BOM section under "Parts". 
I've only stuck to official STL's from Thingiverse by kreagar, but the dxf's for the panels I've taken from two Facebook posts. I think it could probably do with other assorted official parts such as the BMG mount also.

**Comment by plc268** [comment]
> > Adding on, I think it would be useful to collect and organize all the STLs and cad files and post them in one source. There's a whole host of files that are strewn around in separate places and are getting lost in the fold. Not to mention, that thingiverse hasn't exactly been the most reliable host lately.
> 
> Are STL's config or BOM? I figured BOM as it becomes a material :) Anyway I've tried to do that, and it's currently in the BOM section under "Parts".
> I've only stuck to official STL's from Thingiverse by kreagar, but the dxf's for the panels I've taken from two Facebook posts. I think it could probably do with other assorted official parts such as the BMG mount also.

Yea, that's a fine place for it, IMO. I had missed that, and didn't realize that they were already upload. Thanks

**Comment by lostapathy** [comment]
> Are STL's config or BOM? I figured BOM as it becomes a material :) Anyway I've tried to do that, and it's currently in the BOM section under "Parts".
> I've only stuck to official STL's from Thingiverse by kreagar, but the dxf's for the panels I've taken from two Facebook posts. I think it could probably do with other assorted official parts such as the BMG mount also.

I think most importantly, they are not part of the duet config, klipper config, or slicer config - so they don't belong in one of the 3 repos this one will split into.

I think putting other things in the documentation repo is fine for now (since they don't belong here), but we should be mindful of the possibility of (eventually) splitting groups of things into their own repo in the future where it makes sense.

**Comment by natewalck** [comment]
Alright, cut a new repo for slicers and moved the configs over: https://github.com/railcore/slicers

I asked @JohnOCFII to move his configs to a repo under his own account and we'll link to it in the repo.

The next question is, do we want to make Project R3d configs the default? I feel like they should up keep those and we can link to a location of their choosing. I'll ask Joe where the best place is to point people at for Project R3d.

I'd really like this config to be the tested and dogfooded version, but I'd also like to do cool things with it (Like using M98 to include the Z probe configs, thermistors, etc). I think we could break it all out and make it a really clean config. This would also allow us to have multiple configs for a each type of Z probe and people could just comment out the old probe and uncomment the new probe type to swap.

**Comment by dandancheeseandham** [comment]
> The next question is, do we want to make Project R3d configs the default? I feel like they should up keep those and we can link to a location of their choosing. I'll ask Joe where the best place is to point people at for Project R3d.

Yes, I agree. I thought about this myself this morning and came to the conclusion that we should leave Project R3D's files as is, for their kit customers, and just produce our own version.

> I'd really like this config to be the tested and dogfooded version, but I'd also like to do cool things with it (Like using M98 to include the Z probe configs, thermistors, etc). I think we could break it all out and make it a really clean config.

If you'd like a halfway house between these two versions, may I suggest this config?  https://railcore.org/doku.php?id=en:duet_reprap_300zl

I took Kraegars config last October - which I assume has had some decent testing, and then tidied it up manually keeping it in the format and sections (and comments) of the RepRapFirmware generator.
I did try to check for mistakes thoroughly, but as my RC is not built yet, this has not been tested yet. If someone has a spare microSD to blow this config , change a few values to theirs and generally be a guinea pig, that'd be awesome 😄 
I'm fairly happy with it, although I think the Drives section does need verifying as I notice that everyone seems to have slightly different Drive configurations?

My thinking behind this when I did it was that if we can get configurator worked out (which I've been working on here https://railcore.org/doku.php?id=en:300zlconfigcreator ), it will also mean that people can use the configurator (either for their own reasons or because of a major Duet update) and the format will change very little, making comparison and general upkeep easier.

If you're happy with this, I've created a RailcoreII 300ZL folder, with the sys folder with the standard files, a few choice macros and the filament folder (which needs looking at). 
I presume the idea is to be able to download the entire folder onto a microSD card and have a nicely standard and functioning install?

If we are happy with this, perhaps the others should be cleared out also?

**Comment by dandancheeseandham** [comment]
(of course if I'm jumping the gun or otherwise doing something wrong,please do just say so, and we can revert or indeed do whatever - I'm just trying to keep the momentum going 😃  )

**Comment by dandancheeseandham** [comment]
And I won't do any more commits to master. Sorry, still figuring out this Git lark. I think I may have a handle on it now :)

Just thinking about the config.g , are we going for a monolithic file where we ask users to comment or uncomment large sections of the g-code, or should we look at breaking parts out (and making things more modular) using the M98 command?

So for instance, something like

 ;BLTouch - leave commented out if using a IR Probe
 ;M558 P9 X0 Y0 Z1 H5 F50 T6000 A5 S0.02
 ;G31 X2 Y42 Z2.65 P25           ; Customize your offsets appropriately.
 ;M307 H3 A-1 C-1 D-1


could be

;M98 Pbltouch.g          ;BLTouch - leave commented out if using a IR Probe


I have no preference, I just thought I would highlight it.

**Comment by natewalck** [comment]
No worries. For github, you want to make changes on a new branch and then submit a pull request. 

Click **Branch** in the top left, type in what you want the branch name to be and then click "Create branch" to create the branch. Then, edit the files in the UI like usual.

After edits are complete, create a pull request against master and fill out the form. That should do it!

**Comment by lostapathy** [comment]
> So for instance, something like
> 
> ;BLTouch - leave commented out if using a IR Probe
> ;M558 P9 X0 Y0 Z1 H5 F50 T6000 A5 S0.02
> ;G31 X2 Y42 Z2.65 P25 ; Customize your offsets appropriately.
> ;M307 H3 A-1 C-1 D-1
> 
> could be
> ;M98 Pbltouch.g ;BLTouch - leave commented out if using a IR Probe

I think this is the goal - both for me personally and from the discussions that have happened on IRC both in general and with @kraegar.  Picking the correct M98 to uncomment for your configuration is a lot more clear than uncomment blocks of lines, and should be easier for end-users to keep in sync with their machines as well.

I think we need to establish some naming conventions on that, and avoid abbreviations.  i.e, name the files probe-bltouch.g and probe-ir.g rather than ending up with a pir.g

**Comment by natewalck** [comment]
I'm all in favor of spelling it out :) I'll do a pass on this and test on my 250ZL and cut a PR.

**Comment by lostapathy** [comment]
> Alright, cut a new repo for slicers and moved the configs over: https://github.com/railcore/slicers
> 
> I asked @JohnOCFII to move his configs to a repo under his own account and we'll link to it in the repo.
> 
> The next question is, do we want to make Project R3d configs the default? I feel like they should up keep those and we can link to a location of their choosing. I'll ask Joe where the best place is to point people at for Project R3d.
> 
> I'd really like this config to be the tested and dogfooded version, but I'd also like to do cool things with it (Like using M98 to include the Z probe configs, thermistors, etc). I think we could break it all out and make it a really clean config. This would also allow us to have multiple configs for a each type of Z probe and people could just comment out the old probe and uncomment the new probe type to swap.

I agree - the config in github should be as clean and optimized as we (the active community here on github) can make it, and not necessarily what has been shipping with the kits thus far.  I think the strategy you describe will be better for the kits anyway once we get where we want to go.

If Project r3d wants to get involved here to help influence that, they are more than welcome to, but I don't think we're obligated to in any way to base the config here on what they have shipped.  The one place where we might want to defer to the kits in which M98 gets uncommented by default, for cases where there is a clear default with the kits.

If I had to pick one config to start with to optimize from, it'd be the 300ZL config already in the repo, just because that seems like it's seen the most attention out of what we have here.

**Comment by natewalck** [comment]
I talked to Joe and it sounds like they would be willing to use whatever config standard we come up with. They are basically using Tony's 250ZL config with the specific configs for the R3D 300ZL.

I don't see this as a problem. I'll cut a few PRs and then we can discuss.

**Comment by JohnOCFII** [comment]
> I took Kraegars config last October - which I assume has had some decent testing, and then tidied it up manually keeping it in the format and sections (and comments) of the RepRapFirmware generator.

I think this is a good approach.  Let's try to stick with the file format / sections of a "fresh" config from the RepRap configurator.  That way, for folks with other Duet or RRF printers, our format will flow similarly.  I also started with this format in my configs, and just hand edited to match the bits from Tony, etc.

**Comment by lostapathy** [comment]
> > I took Kraegars config last October - which I assume has had some decent testing, and then tidied it up manually keeping it in the format and sections (and comments) of the RepRapFirmware generator.
> 
> I think this is a good approach. Let's try to stick with the file format / sections of a "fresh" config from the RepRap configurator. That way, for folks with other Duet or RRF printers, our format will flow similarly. I also started with this format in my configs, and just hand edited to match the bits from Tony, etc.

I'm guessing we may diverge quite a ways from RepRap configurator as we go deeper down the M98 road.  I actually think long term it'd be really nice to have a railcore-specific configurator (hopefully one that could even update the duet for you automatically)  anyway.

**Comment by lostapathy** [comment]
I'm going to close this issue now - seems like we've done most of the work here and started opening new issues for more specific things.
